### ISSUE TITLE
DiscreteLp.element() more strict on input shape?

### ISSUE THREAD

**Issue by kohr-h** [odlgroup/odl#856]
When switching from flat to multi-dimensional storage, how do we want to deal with flat input to `DiscreteLp.element`? 

While being "hybrid" in the sense that storage was linear but outer appearance was multi-dim. for `DiscreteLpElement`'s, it made sense to support any kind of data with the right size.

Do we want to abandon this? I think it has some potential for bugs, for example
- Why would it make sense to allow input of arbitrary shape that happens to match the size of the space?
  ```python
  >>> space = odl.uniform_discr([0, 0], [1, 1], (2, 12))
  >>> arr = np.ones((2, 3, 4))
  >>> space.element(arr)  # works, WTF??
  ```
- Easy to mess up with C vs. F ordering. It's not simple to get right with explicit reshaping either, but at least you have a chance to get it right.
- Numpy has a stricter policy, sizes must be broadcastable upon assignment, which is kind of what happens here.

My suggestion is to allow only broadcastable shapes as input.

**Comment by adler-j** [comment]
This is related to #568.

**Comment by kohr-h** [comment]
Closed by #1088
